contracts/NFTVesting/LinearVestingNFTTimeLock.sol

```diff
@@ -0,0 +1,181 @@
+     * (in seconds).
+     *
+     *  The discount accumulation during vesting for beneficiary is based off a linear model y = mx
+     *  The developer would have to send ETH to this contract on contract deployement for discount to be applied.
+     */
+    constructor(
+        IERC721 nft_,
+        uint256 tokenId_,
+        address beneficiary_,
+        uint256 vestingStartTime_,
+        uint256 maxDiscountPercentage_,
+        uint256 maxDuration_
+    ) {
+        require(
+            vestingStartTime_ > block.timestamp,
+            "BasicNFTTimelock: vesting start time is before current time"
+        );
+        require(
+            maxDiscountPercentage_ <= 100,
+            "TimeLock: max discount is greater than 100%. Please use a valid maxDiscountPercentage."
+        );
+        require(
+            address(this).balance > 0,
+            "Time:Lock: Eth should be sent to contract before initialization"
+        );
+
+        _nft = nft_;
+        _tokenId = tokenId_;
+        _beneficiary = beneficiary_;
+        _vestingStartTime = vestingStartTime_;
+        _maxDiscountPercentage = maxDiscountPercentage_;
+        _maxDuration = maxDuration_;
+    }
+
+    /**
+     * @dev Returns the smart contract NFT.
+     */
+    function nft() public view virtual returns (IERC721) {
+        return _nft;
+    }
+
+    /**
+     * @dev Returns the token ID of the NFT being held.
+     */
+    function tokenId() public view virtual returns (uint256) {
+        return _tokenId;
+    }
+
+    /**
+     * @dev Returns the beneficiary that will receive the NFT.
+     */
+    function beneficiary() public view virtual returns (address) {
+        return _beneficiary;
+    }
+
+    /**
+     * @dev Returns the time when the NFT are released in seconds since Unix epoch (i.e. Unix timestamp).
+     */
+    function vestingStartTime() public view virtual returns (uint256) {
+        return _vestingStartTime;
+    }
+
+    /**
+     * @dev Returns the max duration that a token is allowed to vest
+     */
+    function maxDuration() public view virtual returns (uint256) {
+        return _maxDuration;
+    }
+
+    /**
+     * @dev Returns the max discount allowed for a token in percentage.
+     */
+    function maxDiscountPercentage() public view virtual returns (uint256) {
+        return _maxDiscountPercentage;
+    }
+
+    /**
+     * @dev Returns duration that NFT has been locked and vesting
+     */
+    function vestedDuration() public view returns (uint256) {
+        return block.timestamp - vestingStartTime();
+    }
+
+    /**
+     * @dev Returns discount percentage for achieved from vesting
+     * Based off: discount = mx
+     */
+    function getDiscountPercentage() public view returns (uint256) {
+        if (block.timestamp < vestingStartTime()) {
+            return 0;
+        }
+        return (maxDiscountPercentage() * vestedDuration()) / maxDuration();
```

Comment:
Prefer grouping of latter 2 variables

Reply:
removed max maxDiscountPercentage to simplify things